megamek/src/megamek/ai/dataset/UnitAction.java

```diff
@@ -27,80 +28,89 @@
  */
 package megamek.ai.dataset;
 
+import java.util.List;
+import java.util.stream.Stream;
+
+import megamek.ai.utility.EntityFeatureUtils;
 import megamek.client.ui.SharedUtility;
-import megamek.common.Coords;
+import megamek.common.Compute;
 import megamek.common.Entity;
+import megamek.common.IAero;
 import megamek.common.MovePath;
 import megamek.common.MoveStep;
-
-import java.util.List;
+import megamek.common.UnitRole;
 
 /**
- * Represents a unit action.
- * @param id unit id
- * @param facing facing direction
- * @param fromX x coordinate of the starting position
- * @param fromY y coordinate of the starting position
- * @param toX x coordinate of the final position
- * @param toY y coordinate of the final position
- * @param hexesMoved number of hexes moved
- * @param distance distance moved
- * @param mpUsed movement points used
- * @param maxMp maximum movement points
- * @param mpP movement points percentage
- * @param heatP heat percentage
- * @param armorP armor percentage
- * @param internalP internal percentage
- * @param jumping jumping
- * @param prone prone
- * @param legal is move legal
+ * UnitAction is a record that represents the action of a unit in the game.
  * @author Luana Coppio
  */
-public record UnitAction(int id, int teamId, int playerId, String chassis, String model, int facing, int fromX, int fromY, int toX, int toY, int hexesMoved, int distance, int mpUsed,
-                         int maxMp, double mpP, double heatP, double armorP, double internalP, boolean jumping, boolean prone,
-                         boolean legal, double chanceOfFailure, List<MovePath.MoveStepType> steps, boolean bot) {
+public record UnitAction(int id, int teamId, int playerId, String chassis, String model, int facing,
+      int fromX, int fromY, int toX, int toY, int hexesMoved, int distance,
+      int mpUsed, int maxMp, double mpP, double heatP, double armorP, double internalP,
+      boolean jumping, boolean prone, boolean legal, double chanceOfFailure,
+      List<MovePath.MoveStepType> steps, boolean bot,
+      boolean hasEcm, int armor, int internal, int bv, int maxRange, int totalDamage,
+      float armorFrontP, float armorLeftP, float armorRightP, float armorBackP, UnitRole role,
+      List<Integer> weaponDamageFacingShortMediumLongRange) {
```

Comment:
While I understand that this is a record, I would highly recommend adding more ergonomic constructors similar to the static method below.
One that accepts just `entity` and `movePath`, similar to the static method, but then populates the values and calls the base constructor, would be useful.

Reply:
Actually, looking at the code again, this would just be extra work for no benefit.
But I'd like to advise that 8 lines of constructor parameters is going to be fragile and prone to breakage.

Reply:
it is a data objet that represents the entire TSV line, there isnt much more to do here other than break into multiple objects, which also would not solve the problem.
Record here is good because it is immutable, and it being prone to breakage, well, I literally can't do better.